src/beast/beast/asio/streambuf.h

```diff
@@ -219,230 +264,417 @@ class basic_streambuf<Allocator>::element
 
-template <class Allocator>
+template<class Allocator>
 auto
-basic_streambuf<Allocator>::mutable_buffers_type::
-    transform::operator() (element const& e) const ->
-        value_type const
+basic_streambuf<Allocator>::mutable_buffers_type::end() const ->
+    const_iterator
 {
-    return value_type (e.data(), &e == &*std::prev(streambuf_->list_.end()) ?
-        streambuf_->out_end_ : e.size()) + (&e == &*streambuf_->out_ ?
-            streambuf_->out_pos_ : 0);
+    return const_iterator{*sb_, sb_->list_.end()};
 }
 
 //------------------------------------------------------------------------------
 
-template <class Allocator>
+template<class Allocator>
 basic_streambuf<Allocator>::~basic_streambuf()
 {
-    for(auto iter = list_.begin(); iter != list_.end();)
-    {
-        auto& e = *iter++;
-        size_type const n = e.alloc_size();
-        alloc_traits::destroy(this->member(), &e);
-        alloc_traits::deallocate(this->member(),
-            reinterpret_cast<char*>(&e), n);
-    }
+    delete_list();
 }
 
-template <class Allocator>
-basic_streambuf<Allocator>::basic_streambuf(std::size_t block_size,
-        Allocator const& alloc)
-    : empty_base_optimization<Allocator>(alloc)
-    , block_size_ (block_size)
-    , block_size_next_ (block_size)
-    , out_ (list_.end())
+template<class Allocator>
+basic_streambuf<Allocator>::basic_streambuf(basic_streambuf&& other)
+    : empty_base_optimization<Allocator>(other.member())
+    , alloc_size_(other.alloc_size_)
+    , in_size_(other.in_size_)
+    , in_pos_(other.in_pos_)
+    , out_pos_(other.out_pos_)
+    , out_end_(other.out_end_)
 {
-    if (! (block_size > 0))
-        throw std::invalid_argument(
-            "basic_streambuf: invalid block_size");
+    static_assert(std::is_copy_constructible<basic_streambuf>::value, "");
+    static_assert(std::is_move_constructible<basic_streambuf>::value, "");
+    static_assert(std::is_copy_assignable<basic_streambuf>::value, "");
+    static_assert(std::is_move_assignable<basic_streambuf>::value, "");
+
+    // VFALCO What about allocator propagate_on_container_move_construct?
+    auto const at_end =
+        other.out_ == other.list_.end();
+    list_ = std::move(other.list_);
+    out_ = at_end ? list_.end() : other.out_;
+    other.in_size_ = 0;
+    other.out_ = other.list_.end();
+    other.in_pos_ = 0;
+    other.out_pos_ = 0;
+    other.out_end_ = 0;
 }
 
-template <class Allocator>
-basic_streambuf<Allocator>::basic_streambuf (basic_streambuf&& other)
-    : empty_base_optimization<Allocator>(other.member())
-    , list_ (std::move(other.list_))
-    , block_size_ (other.block_size_)
-    , block_size_next_ (other.block_size_next_)
-    , in_size_ (other.in_size_)
-    , out_ (other.out_)
-    , in_pos_ (other.in_pos_)
-    , out_pos_ (other.out_pos_)
-    , out_end_ (other.out_end_)
+template<class Allocator>
+auto
+basic_streambuf<Allocator>::operator=(basic_streambuf&& other)
+    -> basic_streambuf&
 {
+    clear();
```

Comment:
Ah...you mean, that on self assignment the thing will end up empty. I'm leaning ever so slightly to checking `if(this == &other)` simply because I don't anticipate move assignments happening at such a high frequency that the cost of a branch prediction could become a limiting factor. And with how convoluted some of these APIs are getting, I could easily see myself doing a self-assign unwittingly.